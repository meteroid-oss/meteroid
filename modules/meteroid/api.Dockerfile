
FROM lukemathwalker/cargo-chef:latest-rust-1-bookworm AS chef
ARG CARGO_REGISTRIES_CRATES_IO_PROTOCOL=sparse
ENV CARGO_NET_GIT_FETCH_WITH_CLI=true
WORKDIR /opt/src

FROM chef AS planner
COPY . .
RUN cargo chef prepare --recipe-path recipe.json

FROM chef AS builder
ARG MOLD_VERSION=1.11.0
ARG PROTO_VERSION=21.8

ARG MOLD_ARCH
ARG PROTO_ARCH
ARG GRPC_HEALTH_PROBE_ARCH

RUN test -n "$MOLD_ARCH" || (echo "MOLD_ARCH not set" && false)
RUN test -n "$PROTO_ARCH" || (echo "PROTO_ARCH not set" && false)
RUN test -n "$GRPC_HEALTH_PROBE_ARCH" || (echo "GRPC_HEALTH_PROBE_ARCH not set" && false)

RUN apt-get update

# install needed packages
RUN DEBIAN_FRONTEND=noninteractive && \
    apt-get -y install --no-install-recommends curl pkg-config unzip build-essential libssl-dev libsasl2-dev openssl cmake clang wget

# Install mold
RUN wget https://github.com/rui314/mold/releases/download/v${MOLD_VERSION}/mold-${MOLD_VERSION}-${MOLD_ARCH}-linux.tar.gz && \
    tar xvfz mold*.tar.gz && \
    mv mold*-linux/bin/* /usr/local/bin && \
    mv mold*-linux/libexec/* /usr/libexec && \
    rm -rf mold*

# Install protoc
RUN wget https://github.com/protocolbuffers/protobuf/releases/download/v${PROTO_VERSION}/protoc-${PROTO_VERSION}-linux-${PROTO_ARCH}.zip && \
    unzip protoc*.zip && \
    mv bin/protoc /usr/local/bin && \
    mv include/google /usr/local/include

# Install grpc health checker
RUN GRPC_HEALTH_PROBE_VERSION=v0.4.24 && \
    wget -qO/bin/grpc_health_probe https://github.com/grpc-ecosystem/grpc-health-probe/releases/download/${GRPC_HEALTH_PROBE_VERSION}/grpc_health_probe-linux-${GRPC_HEALTH_PROBE_ARCH} && \
    chmod +x /bin/grpc_health_probe

# Cleanup
RUN apt-get clean; \
    rm -rf /var/lib/apt/lists/*

COPY --from=planner /opt/src/recipe.json recipe.json

ARG PROFILE
ARG CI
# Build dependencies & cache
RUN cargo chef cook --recipe-path recipe.json --profile $PROFILE --package meteroid
# Build application
COPY . .
RUN cargo build -p meteroid --bin meteroid-api --profile $PROFILE


FROM debian:stable-slim
ARG PROFILE
ARG TARGET_DIR=$PROFILE
RUN apt-get update && \
    apt-get install --no-install-recommends -y ca-certificates libssl3 libsasl2-2 libpq5 && \
    rm -rf /var/lib/apt/lists/*
COPY --from=builder /opt/src/target/$TARGET_DIR/meteroid-api /usr/local/bin/meteroid-api
COPY --from=builder /bin/grpc_health_probe /bin/grpc_health_probe

RUN groupadd --system md --gid 151 \
    && useradd --system --gid md --uid 151 md

USER md

EXPOSE 50061
EXPOSE 8080

CMD ["meteroid-api"]
