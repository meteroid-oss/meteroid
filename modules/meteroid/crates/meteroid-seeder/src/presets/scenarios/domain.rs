use chrono::NaiveDate;
use educe::Educe;
use meteroid_store::StoreResult;
use meteroid_store::domain::{
    BillingMetricAggregateEnum, BillingPeriodEnum, BillingType, CapacityThreshold, DowngradePolicy,
    PlanTypeEnum, SegmentationMatrix, TermRate, UnitConversionRoundingEnum, UpgradePolicy,
    UsagePricingModel,
};
use meteroid_store::errors::StoreError;

#[derive(Clone)]
pub struct Scenario {
    pub name: String, // will be the tenant name
    pub metrics: Vec<BillableMetric>,
    pub plans: Vec<Plan>,
    pub customers: Vec<Customer>,
}

#[derive(Clone)]
pub struct Plan {
    pub name: String,
    pub currency: String,
    pub plan_type: PlanTypeEnum,
    pub components: Vec<PriceComponent>,
}

#[derive(Clone, Educe)]
#[educe(Default)]
pub struct BillableMetric {
    pub code: String,
    pub name: String,
    pub description: Option<String>,
    #[educe(Default = BillingMetricAggregateEnum::Sum)]
    pub aggregation_type: BillingMetricAggregateEnum,
    pub aggregation_key: Option<String>,
    pub unit_conversion_factor: Option<i32>,
    pub unit_conversion_rounding: Option<UnitConversionRoundingEnum>,
    pub segmentation_matrix: Option<SegmentationMatrix>,
    pub usage_group_key: Option<String>,
}

#[derive(Clone)]
pub struct Subscription {
    pub plan_name: String,
    pub start_date: NaiveDate,
}

#[derive(Clone)]
pub struct Customer {
    pub name: String,
    pub email: String,
    pub currency: String,
    pub subscription: Subscription,
}

#[derive(Clone)]
pub enum FeeType {
    Rate {
        rates: Vec<TermRate>,
    },
    Slot {
        rates: Vec<TermRate>,
        slot_unit_name: String,
        upgrade_policy: UpgradePolicy,
        downgrade_policy: DowngradePolicy,
        minimum_count: Option<u32>,
        quota: Option<u32>,
    },
    Capacity {
        metric_code: String,
        thresholds: Vec<CapacityThreshold>,
    },
    Usage {
        metric_code: String,
        pricing: UsagePricingModel,
    },
    ExtraRecurring {
        unit_price: rust_decimal::Decimal,
        quantity: u32,
        billing_type: BillingType,
        cadence: BillingPeriodEnum,
    },
    OneTime {
        unit_price: rust_decimal::Decimal,
        quantity: u32,
    },
}

impl FeeType {
    pub fn metric_code(&self) -> Option<String> {
        match self {
            FeeType::Capacity { metric_code, .. } => Some(metric_code.clone()),
            FeeType::Usage { metric_code, .. } => Some(metric_code.clone()),
            _ => None,
        }
    }
}

#[derive(Clone)]
pub struct PriceComponent {
    pub name: String,
    pub fee: FeeType,
}

impl PriceComponent {
    pub fn to_domain(
        &self,
        metrics: &[meteroid_store::domain::BillableMetric],
    ) -> StoreResult<meteroid_store::domain::PriceComponentNewInternal> {
        let maybe_metric_code = self.fee.metric_code();

        let maybe_metric = maybe_metric_code
            .and_then(|metric_code| metrics.iter().find(|metric| metric.code == metric_code));

        let domain_fee: meteroid_store::domain::FeeType = match &self.fee {
            FeeType::Rate { rates } => meteroid_store::domain::FeeType::Rate {
                rates: rates.clone(),
            },
            FeeType::Slot {
                rates,
                slot_unit_name,
                upgrade_policy,
                downgrade_policy,
                minimum_count,
                quota,
            } => meteroid_store::domain::FeeType::Slot {
                rates: rates.clone(),
                slot_unit_name: slot_unit_name.clone(),
                upgrade_policy: upgrade_policy.clone(),
                downgrade_policy: downgrade_policy.clone(),
                minimum_count: *minimum_count,
                quota: *quota,
            },
            FeeType::Capacity {
                metric_code,
                thresholds,
            } => {
                let metric = maybe_metric.ok_or(StoreError::ValueNotFound(format!(
                    "Metric was not found {}",
                    metric_code
                )))?;
                meteroid_store::domain::FeeType::Capacity {
                    metric_id: metric.id,
                    thresholds: thresholds.clone(),
                }
            }
            FeeType::Usage {
                metric_code,
                pricing,
            } => {
                let metric = maybe_metric.ok_or(StoreError::ValueNotFound(format!(
                    "Metric was not found {}",
                    metric_code
                )))?;
                meteroid_store::domain::FeeType::Usage {
                    metric_id: metric.id,
                    pricing: pricing.clone(),
                }
            }
            FeeType::ExtraRecurring {
                unit_price,
                quantity,
                billing_type,
                cadence,
            } => meteroid_store::domain::FeeType::ExtraRecurring {
                unit_price: *unit_price,
                quantity: *quantity,
                billing_type: billing_type.clone(),
                cadence: cadence.clone(),
            },
            FeeType::OneTime {
                unit_price,
                quantity,
            } => meteroid_store::domain::FeeType::OneTime {
                unit_price: *unit_price,
                quantity: *quantity,
            },
        };

        Ok(meteroid_store::domain::PriceComponentNewInternal {
            name: self.name.clone(),
            fee: domain_fee,
            product_id: None,
        })
    }
}
