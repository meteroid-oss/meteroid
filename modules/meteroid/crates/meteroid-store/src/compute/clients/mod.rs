pub mod slots;
pub mod usage;
