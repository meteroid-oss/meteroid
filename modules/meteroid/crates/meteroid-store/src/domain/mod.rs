pub use api_tokens::*;
pub use billable_metrics::*;
pub use customers::*;
pub use invoices::*;
pub use misc::*;
pub use plans::*;
pub use price_components::*;
pub use product_families::*;
pub use schedules::*;
pub use subscription_components::*;
pub use subscriptions::*;
pub use tenants::*;

pub mod customers;
pub mod invoices;
pub mod plans;

pub mod price_components;
pub mod tenants;

pub mod adjustments;
pub mod api_tokens;
pub mod billable_metrics;
pub mod configs;
pub mod enums;
pub mod misc;
pub mod product_families;
pub mod schedules;
pub mod subscription_components;
pub mod subscriptions;
pub mod users;
pub mod webhooks;
