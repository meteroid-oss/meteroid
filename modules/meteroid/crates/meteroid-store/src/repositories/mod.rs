pub use customers::CustomersInterface;
pub use invoices::InvoiceInterface;
pub use organizations::OrganizationsInterface;
pub use plans::PlansInterface;
pub use product_families::ProductFamilyInterface;
pub use subscriptions::SubscriptionInterface;
pub use tenants::TenantInterface;

pub mod customers;
pub mod invoices;
pub mod plans;
pub mod tenants;

pub mod add_ons;
pub mod api_tokens;
pub mod billable_metrics;
pub mod configs;
pub mod customer_balance;
pub mod historical_rates;
pub mod organizations;
pub mod price_components;
pub mod product_families;
pub mod products;
pub mod schedules;
pub mod stats;
pub mod subscriptions;
pub mod users;
pub mod webhooks;
pub mod invoicing_entities;
mod constants;
