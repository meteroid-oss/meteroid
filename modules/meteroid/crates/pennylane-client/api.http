### List customers/compamies https://pennylane.readme.io/v2.0/reference/getcustomers
GET https://app.pennylane.com/api/external/v2/customers HTTP/1.1
Accept: application/json
Authorization: Bearer {{api_token}}

### Create a company customer https://pennylane.readme.io/v2.0/reference/postcompanycustomer
POST https://app.pennylane.com/api/external/v2/company_customers
Content-Type: application/json
Accept: application/json
Authorization: Bearer {{api_token}}

{
  "name": "first_test_company",
  "billing_address": {
    "address": "Gates 50",
    "postal_code": "50-001",
    "city": "Wroclaw",
    "country_alpha2": "PL"
  },
  "external_reference": "cus_123456789"
}

### Update a company customer https://pennylane.readme.io/v2.0/reference/putcompanycustomer
PUT https://app.pennylane.com/api/external/v2/company_customers/157896015
Content-Type: application/json
Accept: application/json
Authorization: Bearer {{api_token}}

{
  "name": "first_test_company",
  "billing_address": {
    "address": "Gates 50",
    "postal_code": "50-001",
    "city": "Wroclaw",
    "country_alpha2": "PL"
  },
  "external_reference": "cus_123456789"
}

### customer invoices changelog https://pennylane.readme.io/v2.0/docs/tracking-data-changes-with-pennylane-api#available-changelog-endpoints
GET https://app.pennylane.com/api/external/v2/changelogs/customer_invoices HTTP/1.1
Accept: application/json
Authorization: Bearer {{api_token}}

### Upload file https://pennylane.readme.io/v2.0/reference/postfileattachments
POST https://app.pennylane.com/api/external/v2/file_attachments HTTP/1.1
Accept: application/json
Content-Type: multipart/form-data; boundary=MyBoundary
Authorization: Bearer {{api_token}}

--MyBoundary
Content-Disposition: form-data; name="file"; filename="custom_name.pdf"
Content-Type: application/pdf

< ./inv_sample.pdf
--MyBoundary--


### Import customer invoice https://pennylane.readme.io/v2.0/reference/importcustomerinvoices
POST https://app.pennylane.com/api/external/v2/customer_invoices/import HTTP/1.1
Content-Type: application/json
Accept: application/json
Authorization: Bearer {{api_token}}

{
  "file_attachment_id": 1141709,
  "external_reference": "inv_123456789",
  "invoice_number": "PL00001",
  "date": "2025-05-01",
  "deadline": "2025-05-01",
  "customer_id": 157896015,
  "currency": "EUR",
  "currency_amount_before_tax": "200.1",
  "currency_amount": "220.1",
  "amount": "220.1",
  "currency_tax": "20",
  "invoice_lines": [
    {
      "currency_amount": "220.1",
      "amount": "220.1",
      "currency_tax": "20",
      "label": "Test line",
      "quantity": 1.5,
      "raw_currency_unit_price": "20.01",
      "unit": "units",
      "vat_rate": "PL_230"
    }
  ]
}

### Mark customer invoice as paid https://pennylane.readme.io/v2.0/reference/markaspaidcustomerinvoice
PUT https://app.pennylane.com/api/external/v2/customer_invoices/1974109344/mark_as_paid
Authorization: Bearer {{api_token}}

### List customer invoices https://pennylane.readme.io/v2.0/reference/getcustomerinvoices
GET https://app.pennylane.com/api/external/v2/customer_invoices HTTP/1.1
Accept: application/json
Authorization: Bearer {{api_token}}

### Get customer by external_reference https://pennylane.readme.io/v2.0/reference/getcustomers
GET https://app.pennylane.com/api/external/v2/customers?
  filter=[{"field": "external_reference", "operator": "eq", "value": "cus_VTNnzVXRsjJuusdcEPVtl"}] HTTP/1.1
Accept: application/json
Authorization: Bearer {{api_token}}

### Get invoice by external_reference https://pennylane.readme.io/v2.0/reference/getcustomers
GET https://app.pennylane.com/api/external/v2/customer_invoices?
  filter=[{"field": "external_reference", "operator": "eq", "value": "inv_123456789"}] HTTP/1.1
Accept: application/json
Authorization: Bearer {{api_token}}

