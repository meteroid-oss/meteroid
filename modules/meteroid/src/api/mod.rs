mod cors;
pub mod server;

pub mod shared;
pub mod utils;

pub mod addons;
pub mod apitokens;
pub mod bankaccounts;
pub mod billablemetrics;
pub mod connectors;
pub mod coupons;
pub mod customers;
mod domain_mapping;
pub mod errors;
pub mod instance;
pub mod internal;
pub mod invoices;
pub mod invoicingentities;
pub mod organizations;
pub mod plans;
pub mod portal;
pub mod pricecomponents;
pub mod productfamilies;
pub mod productitems;
pub mod schedules;
pub mod sharable;
pub mod stats;
pub mod subscriptions;
pub mod taxes;
pub mod tenants;
pub mod users;
pub mod webhooksout;
