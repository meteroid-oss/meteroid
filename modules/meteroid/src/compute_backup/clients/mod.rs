pub mod subscription;
pub mod usage;
