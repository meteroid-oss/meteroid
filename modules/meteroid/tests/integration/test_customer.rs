use meteroid_grpc::meteroid::api;
use serde_json::json;

use crate::helpers;
use crate::meteroid_it;
use crate::meteroid_it::clients::AllClients;
use crate::meteroid_it::container::{MeteroidSetup, SeedLevel};
use common_domain::ids::{BaseId, ConnectorId, CustomerId, TenantId};
use meteroid_store::domain::ConnectorProviderEnum;
use meteroid_store::repositories::CustomersInterface;
use tonic::Code;

#[tokio::test]
async fn test_customers_basic() {
    // Generic setup
    helpers::init::logging();
    let (_postgres_container, postgres_connection_string) =
        meteroid_it::container::start_postgres().await;
    let setup =
        meteroid_it::container::start_meteroid(postgres_connection_string, SeedLevel::MINIMAL)
            .await;

    let auth = meteroid_it::svc_auth::login(setup.channel.clone()).await;

    let clients = meteroid_it::clients::AllClients::from_channel(
        setup.channel.clone(),
        auth.token.clone().as_str(),
        "TESTORG",
        "testslug",
    );

    let customer_name = "friends and co".to_owned();
    let customer_alias = "fffrrriiieeennndddsss".to_owned();
    let customer_email = "fake@fake.com".to_owned();

    // create customer
    let created = clients
        .customers
        .clone()
        .create_customer(api::customers::v1::CreateCustomerRequest {
            data: Some(api::customers::v1::CustomerNew {
                name: customer_name.to_string(),
                alias: Some(customer_alias.to_string()),
                billing_email: Some(customer_email.to_string()),
                invoicing_emails: Vec::new(),
                phone: None,
                currency: "EUR".to_string(),
                billing_address: None,
                shipping_address: None,
                invoicing_entity_id: None,
                bank_account_id: None,
                vat_number: None,
                custom_vat_rate: None,
            }),
        })
        .await
        .unwrap()
        .into_inner()
        .customer
        .unwrap();

    setup
        .store
        .patch_customer_conn_meta(
            CustomerId::from_proto(created.id.as_str()).unwrap(),
            ConnectorId::new(),
            ConnectorProviderEnum::Hubspot,
            "idk",
        )
        .await
        .unwrap();

    let patched_conn_meta = setup
        .store
        .find_customer_by_id(
            CustomerId::from_proto(created.id.as_str()).unwrap(),
            TenantId::from_proto("018c2c82-3df1-7e84-9e05-6e141d0e751a").unwrap(),
        )
        .await
        .unwrap();

    assert_ne!(patched_conn_meta.conn_meta, None);

    let created_manual = clients
        .customers
        .clone()
        .create_customer(api::customers::v1::CreateCustomerRequest {
            data: Some(api::customers::v1::CustomerNew {
                name: "created_manual".to_string(),
                alias: Some("created_manual".to_string()),
                billing_email: Some("created_manual@meteroid.com".to_string()),
                invoicing_emails: Vec::new(),
                phone: None,
                currency: "EUR".to_string(),
                billing_address: None,
                shipping_address: None,
                invoicing_entity_id: None,
                bank_account_id: None,
                vat_number: None,
                custom_vat_rate: None,
            }),
        })
        .await
        .unwrap()
        .into_inner()
        .customer
        .unwrap();

    assert_eq!(created.name, customer_name.clone());
    assert_eq!(created.alias, Some(customer_alias.clone()));
    assert_eq!(created.billing_email, Some(customer_email.clone()));

    // list by [fake] search
    let list_by_fake = clients
        .customers
        .clone()
        .list_customers(api::customers::v1::ListCustomerRequest {
            search: Some("fake".to_string()),
            sort_by: api::customers::v1::list_customer_request::SortBy::NameAsc as i32,
            pagination: Some(common_grpc::meteroid::common::v1::Pagination {
                per_page: Some(10),
                page: 0,
            }),
        })
        .await
        .unwrap()
        .into_inner()
        .customers;

    assert_eq!(list_by_fake.len(), 0);

    // list by [alias] search
    let list_by_fake = clients
        .customers
        .clone()
        .list_customers(api::customers::v1::ListCustomerRequest {
            search: Some("iiieeennn".to_string()), // part of alias
            sort_by: api::customers::v1::list_customer_request::SortBy::NameAsc as i32,
            pagination: Some(common_grpc::meteroid::common::v1::Pagination {
                per_page: Some(10),
                page: 0,
            }),
        })
        .await
        .unwrap()
        .into_inner()
        .customers;

    assert_eq!(list_by_fake.len(), 1);

    // get by id
    let get_by_id = clients
        .customers
        .clone()
        .get_customer_by_id(api::customers::v1::GetCustomerByIdRequest {
            id: created.id.to_string(),
        })
        .await
        .unwrap()
        .into_inner()
        .customer
        .unwrap();

    assert_eq!(get_by_id.id, created.id.clone());
    assert_eq!(get_by_id.name, customer_name.clone());
    assert_eq!(get_by_id.alias, Some(customer_alias.clone()));
    assert_eq!(get_by_id.currency, "EUR");

    // get by alias
    let get_by_alias = clients
        .customers
        .clone()
        .get_customer_by_alias(api::customers::v1::GetCustomerByAliasRequest {
            alias: customer_alias.to_string(),
        })
        .await
        .unwrap()
        .into_inner()
        .customer
        .unwrap();

    assert_eq!(get_by_alias.id, created.id.clone());
    assert_eq!(get_by_alias.name, customer_name.clone());
    assert_eq!(get_by_alias.alias, Some(customer_alias.clone()));
    assert_eq!(get_by_alias.currency, "EUR");

    // update
    let _ = clients
        .customers
        .clone()
        .update_customer(api::customers::v1::UpdateCustomerRequest {
            customer: Some(api::customers::v1::UpdateCustomer {
                id: created.id.clone(),
                name: Some("new name".to_string()),
                billing_email: None,
                alias: None,
                invoicing_emails: None,
                phone: None,
                balance_value_cents: None,
                currency: None,
                billing_address: None,
                shipping_address: None,
                invoicing_entity_id: None,
                vat_number: None,
                custom_vat_rate: None,
                bank_account_id: None,
            }),
        })
        .await
        .unwrap()
        .into_inner();

    let get_by_id_patched = clients
        .customers
        .clone()
        .get_customer_by_id(api::customers::v1::GetCustomerByIdRequest {
            id: created.id.to_string(),
        })
        .await
        .unwrap()
        .into_inner()
        .customer
        .unwrap();

    assert_eq!(get_by_id_patched.id, created.id.clone());
    assert_eq!(get_by_id_patched.name, "new name".to_string());
    assert_eq!(get_by_id_patched.alias, Some(customer_alias.clone()));
    assert_eq!(get_by_id_patched.currency, "EUR");

    // top up start
    let topped_up = clients
        .customers
        .clone()
        .top_up_customer_balance(api::customers::v1::TopUpCustomerBalanceRequest {
            customer_id: created.id.to_string(),
            cents: 1024,
            notes: Some("test".into()),
        })
        .await
        .unwrap()
        .into_inner()
        .customer
        .unwrap();

    assert_eq!(topped_up.balance_value_cents, 1024);

    let topped_up = clients
        .customers
        .clone()
        .top_up_customer_balance(api::customers::v1::TopUpCustomerBalanceRequest {
            customer_id: created.id.to_string(),
            cents: -1000,
            notes: Some("test".into()),
        })
        .await
        .unwrap()
        .into_inner()
        .customer
        .unwrap();

    assert_eq!(topped_up.balance_value_cents, 24);

    let topped_up_fail = clients
        .customers
        .clone()
        .top_up_customer_balance(api::customers::v1::TopUpCustomerBalanceRequest {
            customer_id: created.id.to_string(),
            cents: -25,
            notes: Some("test".into()),
        })
        .await
        .err()
        .unwrap();

    assert_eq!(topped_up_fail.message(), "negative customer balance");
    assert_eq!(topped_up_fail.code(), Code::FailedPrecondition);

    // top up end

    // bue credits start
    let credits_invoice = clients
        .customers
        .clone()
        .buy_customer_credits(api::customers::v1::BuyCustomerCreditsRequest {
            customer_id: created_manual.id.clone(),
            cents: 1024,
            notes: Some("test".into()),
        })
        .await
        .unwrap()
        .into_inner()
        .invoice
        .unwrap();

    assert_eq!(credits_invoice.total, 1024);
    assert_eq!(credits_invoice.subtotal, 1024);
    assert_eq!(credits_invoice.applied_credits, 0);
    assert_eq!(credits_invoice.customer_id, created_manual.id);
    // bue credits end

    rest_api_test(&setup, &clients).await;
}

async fn rest_api_test(setup: &MeteroidSetup, clients: &AllClients) {
    let api_key = clients
        .api_tokens
        .clone()
        .create_api_token(tonic::Request::new(
            api::apitokens::v1::CreateApiTokenRequest {
                name: "test-api-key".to_string(),
            },
        ))
        .await
        .unwrap()
        .into_inner()
        .api_key;

    let client = reqwest::Client::new();

    // CREATE CUSTOMER
    let mut created = client
        .post(format!(
            "{}/api/v1/customers",
            setup.config.rest_api_external_url
        ))
        .bearer_auth(&api_key)
        .json(&json!({
            "name": "Test Customer REST",
            "alias": "test-customer-rest",
            "billing_email": "billing@meteroid.com",
            "invoicing_emails": ["invoicing@meteroid.com"],
            "phone": "123456789",
            "currency": "EUR",
            "billing_address": {
                "line1": "123 Test St",
                "city": "Test City",
                "zip_code": "12345",
                "country": "Testland"
            },
            "shipping_address": {
               "same_as_billing": true
            },
            "vat_number": "VAT123456",
            "custom_vat_rate": 20,
        }))
        .send()
        .await
        .unwrap()
        .json::<serde_json::Value>()
        .await
        .unwrap();

    let created = scrub_customer_json(
        &mut created,
        &["id", "invoicing_entity_id", "bank_account_id"],
    );

    insta::assert_json_snapshot!("rest_created", &created);

    // GET BY ALIAS
    let mut get_by_alias = client
        .get(format!(
            "{}/api/v1/customers/test-customer-rest",
            setup.config.rest_api_external_url
        ))
        .bearer_auth(&api_key)
        .send()
        .await
        .unwrap()
        .json::<serde_json::Value>()
        .await
        .unwrap();

    let invoicing_entity_id = get_by_alias
        .get("invoicing_entity_id")
        .cloned()
        .unwrap()
        .as_str()
        .unwrap()
        .to_string();

    let get_by_alias = scrub_customer_json(
        &mut get_by_alias,
        &["id", "invoicing_entity_id", "bank_account_id"],
    );

    insta::assert_json_snapshot!("rest_get_by_alias", &get_by_alias);

    // LIST CUSTOMERS
    let mut list = client
        .get(format!(
            "{}/api/v1/customers",
            setup.config.rest_api_external_url
        ))
        .bearer_auth(&api_key)
        .send()
        .await
        .unwrap()
        .json::<serde_json::Value>()
        .await
        .unwrap();

    let list = scrub_customer_json(&mut list, &["id", "invoicing_entity_id", "bank_account_id"]);
    insta::assert_json_snapshot!("rest_list", &list);

    // UPDATE CUSTOMER
    let mut updated = client
        .put(format!(
            "{}/api/v1/customers/test-customer-rest",
            setup.config.rest_api_external_url
        ))
        .bearer_auth(&api_key)
        .json(&json!({
            "name": "Test Customer REST",
            "alias": "test-customer-rest",
            "billing_email": "billing@meteroid.com",
            "invoicing_emails": ["invoicing@meteroid.com"],
            "phone": "123456789",
            "currency": "USD",
            "billing_address": {
                "line1": "123 Test St",
                "city": "Test City",
                "zip_code": "12345",
                "country": "Testland"
            },
            "shipping_address": {
               "same_as_billing": true
            },
            "vat_number": "VAT123456",
            "custom_vat_rate": 20,
            "invoicing_entity_id": invoicing_entity_id,
        }))
        .send()
        .await
        .unwrap()
        .json::<serde_json::Value>()
        .await
        .unwrap();

    let updated = scrub_customer_json(
        &mut updated,
        &["id", "invoicing_entity_id", "bank_account_id"],
    );
    insta::assert_json_snapshot!("rest_updated", &updated);

    // DELETE CUSTOMER
    let delete_response = client
        .delete(format!(
            "{}/api/v1/customers/test-customer-rest",
            setup.config.rest_api_external_url
        ))
        .bearer_auth(&api_key)
        .send()
        .await
        .unwrap();

    assert_eq!(delete_response.status(), reqwest::StatusCode::NO_CONTENT);
}

/// Mask non-static fields like ids/timestamps in JSON values.
fn scrub_customer_json(value: &mut serde_json::Value, ids: &[&'static str]) -> serde_json::Value {
    match value {
        serde_json::Value::Object(obj) => {
            for (k, v) in obj.iter_mut() {
                if ids.contains(&k.as_str()) {
                    if let Some(id_str) = v.as_str() {
                        if let Some(pos) = id_str.rfind('_') {
                            let obfuscated = format!("{}{}", &id_str[..=pos], "xxx");
                            *v = serde_json::Value::String(obfuscated);
                        }
                    }
                } else {
                    *v = scrub_customer_json(&mut v.take(), ids);
                }
            }
        }
        serde_json::Value::Array(arr) => {
            for v in arr.iter_mut() {
                *v = scrub_customer_json(&mut v.take(), ids);
            }
        }
        _ => {}
    }
    value.clone()
}
