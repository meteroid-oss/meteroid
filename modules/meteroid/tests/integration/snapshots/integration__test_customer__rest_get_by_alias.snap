---
source: modules/meteroid/tests/integration/test_customer.rs
assertion_line: 376
expression: "&get_by_alias"
---
{
  "alias": "test-customer-rest",
  "bank_account_id": null,
  "billing_address": {
    "city": "Test City",
    "country": "Testland",
    "line1": "123 Test St",
    "zip_code": "12345"
  },
  "billing_email": "billing@meteroid.com",
  "currency": "EUR",
  "custom_vat_rate": 20,
  "id": "cus_xxx",
  "invoicing_emails": [
    "invoicing@meteroid.com"
  ],
  "invoicing_entity_id": "ive_xxx",
  "name": "Test Customer REST",
  "phone": "123456789",
  "shipping_address": {
    "same_as_billing": true
  },
  "vat_number": "VAT123456"
}
